src/persist-client/src/internal/apply.rs

```diff
@@ -83,68 +88,145 @@ where
             .await?;
-        // TODO: StateCache usage would look like the following.
-        if false {
-            let _state = shared_states
-                .get::<K, V, T, D, _, _>(shard_id, || {
-                    metrics
-                        .cmds
-                        .init_state
-                        .run_cmd(&shard_metrics, |_cas_mismatch_metric| {
-                            // No cas_mismatch retries because we just use the returned
-                            // state on a mismatch.
-                            state_versions.maybe_init_shard(&shard_metrics)
-                        })
-                })
-                .await?;
-        }
         Ok(Applier {
             cfg,
             metrics,
             shard_metrics,
             state_versions,
+            shard_id,
             state,
         })
     }
 
-    // TODO: Remove usages of this.
-    pub fn state(&self) -> &TypedState<K, V, T, D> {
-        &self.state
+    pub async fn fetch_upper(&mut self) -> Antichain<T> {
+        self.fetch_and_update_state(None).await;
+        self.upper()
+    }
+
+    /// A point-in-time read of `upper` from the current state.
+    ///
+    /// Due to sharing state with other handles, successive reads to this fn or any other may
+    /// see a different version of state, even if this Applier has not explicitly fetched and
+    /// updated to the latest state.
+    pub fn upper(&self) -> Antichain<T> {
+        self.state
+            .read_lock(&self.metrics.locks.applier_read_cacheable, |state| {
+                state.upper().clone()
+            })
+    }
+
+    /// A point-in-time read of `since` from the current state.
+    ///
+    /// Due to sharing state with other handles, successive reads to this fn or any other may
+    /// see a different version of state, even if this Applier has not explicitly fetched and
+    /// updated to the latest state.
+    #[cfg(test)]
+    pub fn since(&self) -> Antichain<T> {
+        self.state
+            .read_lock(&self.metrics.locks.applier_read_cacheable, |state| {
+                state.since().clone()
+            })
+    }
+
+    /// A point-in-time read of `seqno` from the current state.
+    ///
+    /// Due to sharing state with other handles, successive reads to this fn or any other may
+    /// see a different version of state, even if this Applier has not explicitly fetched and
+    /// updated to the latest state.
+    pub fn seqno(&self) -> SeqNo {
+        self.state
+            .read_lock(&self.metrics.locks.applier_read_cacheable, |state| {
+                state.seqno
+            })
+    }
+
+    /// A point-in-time read of `is_tombstone` from the current state.
+    ///
+    /// Due to sharing state with other handles, successive reads to this fn or any other may
+    /// see a different version of state, even if this Applier has not explicitly fetched and
+    /// updated to the latest state.
+    pub fn is_tombstone(&self) -> bool {
```

Comment:
and this one can guarantee that once it returns true it will always return true after?